mne/preprocessing/maxwell.py

```diff
@@ -0,0 +1,492 @@
+    """Convert spherical partial derivatives to cartesian coords.
+
+    Note: Because we are dealing with partial derivatives, this calculation is
+    not a static transformation. The transformation matrix itself is dependent
+    on azimuth and polar coord.
+    See mathworld.wolfram.com/SphericalCoordinates.html Eq. 96
+
+    Parameters
+    ----------
+    sph_pts : ndarray, shape (n_points, 3)
+        Array containing spherical coordinates points (rad, azimuth, polar)
+    sph_grads : ndarray, shape (n_points, 3)
+        Array containing partial derivatives at each spherical coordinate
+
+    Returns
+    -------
+    cart_grads : ndarray, shape (n_points, 3)
+        Array containing partial derivatives in Cartesian coordinates (x, y, z)
+    """
+
+    cart_grads = np.zeros_like(sph_grads)
+
+    # TODO: needs vectorization, currently matching Jussi's code for debugging
+    for pt_i, (sph_pt, sph_grad) in enumerate(zip(sph_pts, sph_grads)):
+        # get cosine and sine of azimuth and polar coord
+        c_a, s_a = np.cos(sph_pt[1]), np.sin(sph_pt[1])
+        c_p, s_p = np.cos(sph_pt[2]), np.sin(sph_pt[2])
+
+        trans = np.array([[c_a * s_p, -s_a, c_a * c_p],
+                          [s_a * s_p, c_a, c_p * s_a],
+                          [c_p, 0, -s_p]])
+
+        cart_grads[pt_i, :] = np.dot(trans, sph_grad)
+
+    return cart_grads
+
+
+def _cart_to_sph(cart_pts):
+    """Convert Cartesian coordinates to spherical coordinates
+
+    Parameters
+    ----------
+    cart_pts : ndarray, shape (n_points, 3)
+        Array containing points in Cartesian coordinates (x, y, z)
+
+    Returns
+    -------
+    ndarray, shape (n_points, 3)
+        Array containing points in spherical coordinates (rad, azimuth, polar)
+    """
+
+    rad = np.linalg.norm(cart_pts, axis=1)
+    az = np.arctan2(cart_pts[:, 1], cart_pts[:, 0])
+    pol = np.arccos(cart_pts[:, 2] / rad)
+
+    return np.c_[rad, az, pol]
+
+
+def _make_coils(info, accurate=True, coil_def=None):
+    """Prepare dict of MEG coils and their information
+
+    Parameters
+    ----------
+    info : instance of mne.io.meas_info.Info | str
+        If str, then it should be a filename to a Raw, Epochs, or Evoked
+        file with measurement information. If dict, should be an info
+        dict (such as one from Raw, Epochs, or Evoked).
+    coil_def : str | None
+        Filepath to the coil definitions file.
+    accuracy : bool
+        Accuracy of coil information.
+
+    Returns
+    -------
+    megcoils, meg_info : dict
+        MEG coils and information dict
+    """
+
+    if accurate:
+        accuracy = FIFF.FWD_COIL_ACCURACY_ACCURATE
+    else:
+        accuracy = FIFF.FWD_COIL_ACCURACY_NORMAL
+    info_extra = 'info'
+    meg_info = None
+    megnames, megcoils, compcoils = [], [], []
+
+    # MEG channels
+    # TODO: Should we exclude bads (in info['bads'])
+    picks = pick_types(info, meg=True, eeg=False, ref_meg=False,
+                       exclude=[])
```

Comment:
We want to exclude bad channels from being used to construct the multipolar activations, but we want to include them in the reconstructed data (and clear `info['bads']` since the reconstructed / filtered data has effectively been interpolated).